File: src/org/elixir_lang/psi/scope/call_definition_clause/Variants.java
package org.elixir_lang.psi.scope.call_definition_clause;

import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Key;
import com.intellij.psi.PsiElement;
import com.intellij.psi.ResolveState;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StubIndex;
import com.intellij.psi.util.PsiTreeUtil;
import gnu.trove.THashMap;
import org.elixir_lang.annonator.Parameter;
import org.elixir_lang.psi.NamedElement;
import org.elixir_lang.psi.call.Call;
import org.elixir_lang.psi.call.Named;
import org.elixir_lang.psi.scope.CallDefinitionClause;
import org.elixir_lang.psi.stub.index.AllName;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.ENTRANCE;

public class Variants extends CallDefinitionClause {
    /*
     * CONSTANTS
     */

    private static final Key<Call> ENTRANCE_CALL_DEFINITION_CLAUSE = new Key<Call>("ENTRANCE_CALL_DEFINITION_CLAUSE");

    /*
     * Static Methods
     */

    @Nullable
    public static List<LookupElement> lookupElementList(@NotNull PsiElement entrance) {
        Variants variants = new Variants();

        Parameter parameter = Parameter.putParameterized(new Parameter(entrance));
        Call entranceCallDefinitionClause = null;

        if (parameter.isCallDefinitionClauseName()) {
            entranceCallDefinitionClause = (Call) parameter.parameterized;
        }

        ResolveState resolveState = ResolveState
                .initial()
                .put(ENTRANCE, entrance)
                .put(ENTRANCE_CALL_DEFINITION_CLAUSE, entranceCallDefinitionClause);

        PsiTreeUtil.treeWalkUp(
                variants,
                entrance,
                entrance.getContainingFile(),
                resolveState
        );
        List<LookupElement> lookupElementList = new ArrayList<LookupElement>();
        lookupElementList.addAll(variants.getLookupElementCollection());
        variants.addProjectNameElementsTo(lookupElementList, entrance, entranceCallDefinitionClause);

        return lookupElementList;
    }

    /*
     * Fields
     */

    @Nullable
    private Map<PsiElement, LookupElement> lookupElementByPsiElement = null;

    /*
     * Protected Instance Methods
     */

    /**
     * Called on every {@link Call} where {@link org.elixir_lang.structure_view.element.CallDefinitionClause#is} is
     * {@code true} when checking tree with {@link #execute(Call, ResolveState)}
     *
     * @param element
     * @param state
     * @return {@code true} to keep searching up tree; {@code false} to stop searching.
     */
    @Override
    protected boolean executeOnCallDefinitionClause(Call element, ResolveState state) {
        Call entranceCallDefinitionClause = state.get(ENTRANCE_CALL_DEFINITION_CLAUSE);

        if (entranceCallDefinitionClause == null || !element.isEquivalentTo(entranceCallDefinitionClause)) {
            addToLookupElementByPsiElement(element);
        }

        return true;
    }

    /**
     * Whether to continue searching after each Module's children have been searched.
     *
     * @return {@code true} to keep searching up the PSI tree; {@code false} to stop searching.
     */
    @Override
    protected boolean keepProcessing() {
        return false;
    }

    /*
     * Private Instance Methods
     */

    private void addToLookupElementByPsiElement(@NotNull Call call) {
        if (call instanceof Named) {
            Named named = (Named) call;

            PsiElement nameIdentifier = named.getNameIdentifier();

            if (nameIdentifier != null) {
                if (lookupElementByPsiElement == null || !lookupElementByPsiElement.containsKey(nameIdentifier)) {
                    if (lookupElementByPsiElement == null) {
                        lookupElementByPsiElement = new THashMap<PsiElement, LookupElement>();
                    }

                    String name = nameIdentifier.getText();

                    lookupElementByPsiElement.put(
                            nameIdentifier,
                            LookupElementBuilder.createWithSmartPointer(
                                    name,
                                    nameIdentifier
                            ).withRenderer(
                                    new org.elixir_lang.code_insight.lookup.element_renderer.CallDefinitionClause(name)
                            )
                    );
                }
            }
        }
    }

    private void addProjectNameElementsTo(@NotNull List<LookupElement> lookupElementList,
                                          @NotNull PsiElement entrance,
                                          @Nullable Call entranceCallDefinitionClause) {
        Project project = entrance.getProject();
        /* getAllKeys is not the actual keys in the actual project.  They need to be checked.
           See https://intellij-support.jetbrains.com/hc/en-us/community/posts/207930789-StubIndex-persisting-between-test-runs-leading-to-incorrect-completions */
        Collection<String> indexedNameCollection = StubIndex.getInstance().getAllKeys(AllName.KEY, project);
        GlobalSearchScope scope = GlobalSearchScope.allScope(project);

        for (String indexedName : indexedNameCollection) {
            Collection<NamedElement> indexedNameNamedElementCollection = StubIndex.getElements(
                    AllName.KEY,
                    indexedName,
                    project,
                    scope,
                    NamedElement.class
            );

            for (NamedElement indexedNameNamedElement : indexedNameNamedElementCollection) {
                if (entranceCallDefinitionClause == null ||
                        !indexedNameNamedElement.isEquivalentTo(entranceCallDefinitionClause)) {
                    lookupElementList.add(
                            org.elixir_lang.code_insight.lookup.element.CallDefinitionClause.createWithSmartPointer(
                                    indexedName,
                                    indexedNameNamedElement
                            )
                    );
                }
            }
        }
    }

    @NotNull
    private Collection<LookupElement> getLookupElementCollection() {
        Collection<LookupElement> lookupElementCollection;

        if (lookupElementByPsiElement != null) {
            lookupElementCollection = lookupElementByPsiElement.values();
        } else {
            lookupElementCollection = Collections.emptySet();
        }

        return lookupElementCollection;
    }
}

File: src/org/elixir_lang/psi/scope/call_definition_clause/Variants.kt
package org.elixir_lang.psi.scope.call_definition_clause

import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.util.Key
import com.intellij.psi.PsiElement
import com.intellij.psi.ResolveState
import com.intellij.psi.util.PsiTreeUtil
import org.elixir_lang.annotator.Parameter
import org.elixir_lang.psi.ElixirAtom
import org.elixir_lang.psi.ElixirIdentifier
import org.elixir_lang.psi.call.Call
import org.elixir_lang.psi.impl.ElixirPsiImplUtil.ENTRANCE
import org.elixir_lang.psi.impl.call.finalArguments
import org.elixir_lang.psi.impl.stripAccessExpression
import org.elixir_lang.psi.putInitialVisitedElement
import org.elixir_lang.psi.putVisitedElement
import org.elixir_lang.psi.scope.CallDefinitionClause
import org.elixir_lang.semantic.call.definition.Delegation

class Variants : CallDefinitionClause() {
    private var lookupElementByPsiElementName: MutableMap<Pair<PsiElement, String>, LookupElement> = mutableMapOf()

    private val lookupElementCollection: Collection<LookupElement>
        get() = lookupElementByPsiElementName.values

    override fun execute(definition: org.elixir_lang.semantic.call.Definition, state: ResolveState): Boolean {
        definition.nameArityInterval?.let { nameArityInterval ->
            val name = nameArityInterval.name

            definition.clauses.forEach { clause ->
                val psiElement = clause.psiElement

                lookupElementByPsiElementName.computeIfAbsent(psiElement to name) { (element, name) ->
                    LookupElementBuilder.createWithSmartPointer(
                        name,
                        element
                    ).withRenderer(
                        org.elixir_lang.code_insight.lookup.element_renderer.CallDefinitionClause(name)
                    )
                }
            }
        }

        return true
    }

    override fun execute(delegation: Delegation, state: ResolveState): Boolean {
        delegation.name?.let { name ->
            val psiElement = delegation.psiElement

            lookupElementByPsiElementName.computeIfAbsent(psiElement to name) { (_, name) ->
                LookupElementBuilder.createWithSmartPointer(
                    name,
                    psiElement
                ).withRenderer(
                    org.elixir_lang.code_insight.lookup.element_renderer.Delegation(name)
                )
            }
        }

        return true
    }

    override fun executeOnEExFunctionFrom(element: Call, state: ResolveState): Boolean {
        element.finalArguments()?.let { arguments ->
            arguments[1].stripAccessExpression().let { it as? ElixirAtom }?.node?.lastChildNode?.text?.let { name ->
                lookupElementByPsiElementName.computeIfAbsent(element to name) { (_, name) ->
                    LookupElementBuilder.createWithSmartPointer(
                        name,
                        element
                    ).withRenderer(
                        org.elixir_lang.code_insight.lookup.element_renderer.EExFunctionFrom(name)
                    )
                }
            }
        }

        return true
    }

    override fun executeOnMixGeneratorEmbed(element: Call, state: ResolveState): Boolean {
        element.finalArguments()?.first()?.stripAccessExpression()
            ?.let { it as? ElixirAtom }?.node?.lastChildNode?.text?.let { prefix ->
                val suffix = element.functionName()!!.removePrefix("embed_")
                val name = "${prefix}_${suffix}"

                lookupElementByPsiElementName.computeIfAbsent(element to name) { (element, name) ->
                    val renderer = when (suffix) {
                        "template" -> org.elixir_lang.code_insight.lookup.element_renderer.mix.generator.EmbedTemplate(
                            name
                        )
                        "text" -> org.elixir_lang.code_insight.lookup.element_renderer.mix.generator.EmbedText(name)
                        else -> TODO()
                    }

                    LookupElementBuilder
                        .createWithSmartPointer(name, element)
                        .withRenderer(renderer)
                }
            }

        return true
    }

    /**
     * Whether to continue searching after each Module's children have been searched.
     *
     * @return `true` to keep searching up the PSI tree; `false` to stop searching.
     */
    override fun keepProcessing(): Boolean = false


    companion object {
        private val ENTRANCE_CALL_DEFINITION_CLAUSE = Key<Call>("ENTRANCE_CALL_DEFINITION_CLAUSE")

        @JvmStatic
        fun lookupElementList(entrance: Call): List<LookupElement> {
            val parameter = Parameter.putParameterized(Parameter(entrance))
            val entranceCallDefinitionClause: Call? = if (parameter.isCallDefinitionClauseName) {
                parameter.parameterized as Call?
            } else {
                null
            }

            return lookupElementList(entrance, entranceCallDefinitionClause)
        }

        @JvmStatic
        fun lookupElementList(entrance: ElixirIdentifier): List<LookupElement> = lookupElementList(entrance, null)

        private fun lookupElementList(entrance: PsiElement, entranceCallDefinitionClause: Call?): List<LookupElement> {
            val variants = Variants()

            val resolveState = ResolveState
                .initial()
                .put(ENTRANCE, entrance)
                .put(ENTRANCE_CALL_DEFINITION_CLAUSE, entranceCallDefinitionClause)
                .putInitialVisitedElement(entrance)

            if (entranceCallDefinitionClause != null) {
                resolveState.putVisitedElement(entranceCallDefinitionClause)
            }

            PsiTreeUtil.treeWalkUp(
                variants,
                entrance,
                entrance.containingFile,
                resolveState
            )
            val lookupElementList = ArrayList<LookupElement>()
            lookupElementList.addAll(variants.lookupElementCollection)

            return lookupElementList
        }
    }
}
